# This script tries to narrow down the potential TDMD targets to a more manageable list
# It takes as input the large list of candidate UTR and ncRNA sites for a single miRNA (generated by TDMD_trigger_finder), 
# And parses that list for top candidates based on loop length

##############################################################################
library(tidyr)
library(dplyr)
##############################################################################
# First import and combine the 3'UTR and ncRNA data #
##############################################################################
# Import the UTR and ncRNA potential trigger sites for a given miRNA
pc.target.list <- read.table("3'UTR trigger list path", sep = ',', header = TRUE)
colnames(pc.target.list)[1] <- 'gene'
nc.target.list <- read.table("ncRNA trigger list path", sep = ',', header = TRUE)
colnames(nc.target.list)[1] <- 'gene'

# Add in information about RPM values (or TPM values if preferred) for potential triggers
# To use this code, the user will need to supply information from RNAsequencing
# The table should be two columns, first colummn is the gene name, second column is the RPM (or TPM value)
RPM.info <- read.table("RPM info file path", sep = '\t', header = TRUE)
colnames(RPM.info)[1] <- 'gene'
pc.targets <- merge(RPM.info[,c(1,3)], pc.target.list, by = 'gene')
nc.targets <- merge(RPM.info[,c(1,3)], nc.target.list, by = 'gene')

# Then generate a final combined list of potential trigger targets (removing any duplicate entries)
all.targets <- rbind(pc.targets[,c(1,2,3,4,5)], nc.targets[,c(1,2,3,4,5)])
colnames(all.targets) <- c('gene', 'RPM', 'sequence', 'threep.end', 'seed.start')
unique.gene.targets <- all.targets[!duplicated(all.targets$sequence),]

##############################################################################
# Then select candidate sites based on length of the loop between the regions of 3' and seed complementarity
##############################################################################
# Currently this code allows the offset of the loop to be anywhere from -4 to 10 nucleotides (offset = # nucleotides on trigger side of loop - # nucleoties on miRNA side)
bulge.selected.list <- unique.gene.targets[between(unique.gene.targets$seed.start, 1, 15),]

# This file is then written out, and then fed directly into 3_trigger_miRNA_duplex_energy.py to determine the duplex stability for the site
write.csv(bulge.selected.list, "~/Documents/Bartel/ZSwim8/DME_TDMD_sites/miR311/select_dme-miR-311-3p.csv", row.names = FALSE, quote = FALSE)